packages/adapter-components/src/definitions/user/user_config.ts

```diff
@@ -124,3 +129,97 @@ export const adapterConfigFromConfig = <
+    const transformationConfig = defs.transformation
+    if (!transformationConfig) {
+      return undefined
+    }
+    const updatedElemIDConfig: ElemIDCustomization<TCustomNameMappingOptions> = {}
+    const { idFields, extendsParentId, nameMapping } = transformationConfig
+    if (extendsParentId !== undefined) {
+      updatedElemIDConfig.extendsParent = extendsParentId
+      delete transformationConfig.extendsParentId
+    }
+    if (idFields !== undefined) {
+      const updatedIdFields = idFields.map(fieldName =>
+        isReferencedIdField(fieldName)
+          ? { fieldName: dereferenceFieldName(fieldName), isReference: true }
+          : { fieldName },
+      )
+      updatedElemIDConfig.parts = updatedIdFields
+      delete transformationConfig.idFields
+    }
+    if (nameMapping !== undefined) {
+      if (updatedElemIDConfig.parts) {
+        updatedElemIDConfig.parts.forEach(part => _.set(part, 'mapping', nameMapping))
+        delete transformationConfig.nameMapping
+      } else {
+        // handaling the case of when mappings exists but idFields are missing is not trivial, because we don't know which parts should be updated
+        log.warn('found nameMapping without idFields for type %s', typeName)
+      }
+    }
+
+    log.debug(
+      'converted elemIDs definitions for type %s from %o to %o',
+      typeName,
+      _.pick(transformationConfig, ['idFields', 'extendsParentId', 'nameMapping']),
+      updatedElemIDConfig,
+    )
+    return updatedElemIDConfig
+  })
+
+  const typesWithIds = _.omitBy(convertedIds, _.isEmpty)
+
+  return _.isEmpty(typesWithIds) ? {} : { elemID: typesWithIds }
+}
+
+export const updateDeprecatedConfig = (
+  config: InstanceElement,
+): { config: InstanceElement; message: string } | undefined => {
+  const apiDefs = config.value?.fetch?.apiDefinitions
+  if (!apiDefs) {
+    return undefined
+  }
+
+  log.warn('adapter config contains deprecated api definitions: %s', safeJsonStringify(apiDefs))
+  const updatedConfig = config.clone()
+  const updatedElemIDs = updateElemIDDefinitions(updatedConfig.value.fetch.apiDefinitions)
+  if (_.isEmpty(updatedElemIDs)) {
+    log.debug('found no elemID definitions to update in config')
+    return undefined
+  }
+  
+  const cleanedApiDefs = objects.cleanEmptyObjects(updatedConfig.value?.fetch?.apiDefinitions)
+  updatedConfig.value = {
+    ...updatedConfig.value,
+    fetch: {
+      ...updatedConfig.value.fetch,
+      ...updatedElemIDs,
+      apiDefinitions: cleanedApiDefs ? { ...cleanedApiDefs } : undefined,
+    },
+  }
+
+  if (updatedConfig.value.fetch.apiDefinitions) {
+    log.error('some deprecated apiDefinitions remained in config: %s', safeJsonStringify(updatedConfig.value.fetch.apiDefinitions))
+  }
+  return {
+    config: updatedConfig,
+    message:
+      'The configuration options "apiDefinitions" is deprecated. The following changes will update the deprecated options to the "fetch" configuration option.',
```

Comment:
nit: maybe something like 
```suggestion
      'Elem ID customizations are now under `fetch.elemID`. The following changes will upgrade the deprecated definitions from `apiDefinitions` to the new location.',
```
(not sure if we can use the backticks here or not, if not we can quote)